### PR TITLE
8319324: FFM: Reformat javadocs

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to reformat all the JavaDocs for the FFM API. This would bring the FFM API docs more in line with the existing Java documentation (see below).  Occasional drive-by fixes are also included in this PR (such as spelling and capitalization).

I am aware this PR will (if approved) make a significant mark in the change logs which is regrettable.

Background:

Older classes like `Object` and `List` have a maximum line length of 80 characters whereas newer classes like `ScopedValue` have a maximum line length of 90 characters.

The FFM API currently has javadoc lines that exceed 135 characters per line. It is also customary to use double spaces when starting a new sentence.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319324](https://bugs.openjdk.org/browse/JDK-8319324): FFM: Reformat javadocs (**Sub-task** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16518/head:pull/16518` \
`$ git checkout pull/16518`

Update a local copy of the PR: \
`$ git checkout pull/16518` \
`$ git pull https://git.openjdk.org/jdk.git pull/16518/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16518`

View PR using the GUI difftool: \
`$ git pr show -t 16518`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16518.diff">https://git.openjdk.org/jdk/pull/16518.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16518#issuecomment-1794244037)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 1]
```diff
@@ -1377,360 +1590,410 @@ static void copy(MemorySegment srcSegment, ValueLayout srcElementLayout, long sr
...
      * @throws IndexOutOfBoundsException if {@code offset > byteSize() - layout.byteSize()}
      */
     boolean get(ValueLayout.OfBoolean layout, long offset);
 
     /**
      * Writes a boolean into this segment at the given offset, with the given layout.
      *
-     * @param layout the layout of the region of memory to be written.
-     * @param offset offset in bytes (relative to this segment address) at which this access operation will occur.
+     * @param offset offset in bytes (relative to this segment address) at which this
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1630]: The first `@param` (layout) here has been deleted?

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1630]: Not just here - happened elsewhere too

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 2]
```diff
@@ -1377,360 +1590,410 @@ static void copy(MemorySegment srcSegment, ValueLayout srcElementLayout, long sr
...
      * @throws WrongThreadException if this method is called from a thread {@code T},
      *         such that {@code isAccessibleBy(T) == false}
      * @throws IllegalArgumentException if the access operation is
-     *         <a href="MemorySegment.html#segment-alignment">incompatible with the alignment constraint</a> in the provided layout.
+     *         <a href="MemorySegment.html#segment-alignment">incompatible with the alignment constraint</a>
+     *         in the provided layout
      * @throws IndexOutOfBoundsException if {@code offset > byteSize() - layout.byteSize()}
-     * @throws UnsupportedOperationException if this segment is {@linkplain #isReadOnly() read-only}
-     * @throws UnsupportedOperationException if {@code value} is not a {@linkplain #isNative() native} segment
+     * @throws UnsupportedOperationException if this segment is
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1910]: one `UOE` went missing

**Review by mcimadamore — COMMENTED** [review]
Generally looks good (I've been relying a lot on github diff annotations, so I hope those are accurate :-) ). There seem to be a couple of issues with deleted text.

**Review by mcimadamore — COMMENTED** [review]
You might also want to look into the MethodHandles XYZCoordinates combinators (as they have also been added as part of FFM). And, also, `ModuleLayer::Controller::enableNativeAccess`, `Module::isNativeAccessEnabled` and `FileChannel::map(... Arena)`.

**Comment by JornVernee** [comment]
FWIW, I've been sticking to a 120 character limit per line when writing javadoc. Why should we use 90 columns specifically?

> It is also customary to use double spaces when starting a new sentence.

I thought this was the exception rather than the rule, to be honest.

**Comment by minborg** [comment]
> FWIW, I've been sticking to a soft 120 character limit per line when writing javadoc. Why should we use 90 columns specifically?
> 
> > It is also customary to use double spaces when starting a new sentence.
> 
> I thought this was the exception rather than the rule, to be honest.

We should try to get some clarity on this. I've sampled a handful of prominent Java classes and they have the style as in this PR. If we can find support that other styles are to be preferred, I am happy to update accordingly.

File: src/java.base/share/classes/java/lang/foreign/Arena.java [hunk 3]
```diff
@@ -226,45 +240,49 @@ class Holder {
 
     /**
      * {@return a new confined arena} Segments allocated with the confined arena can be
-     * {@linkplain MemorySegment#isAccessibleBy(Thread) accessed} by the thread that created the arena,
-     * the arena's <em>owner thread</em>.
+     *          {@linkplain MemorySegment#isAccessibleBy(Thread) accessed} by the thread
+     *          that created the arena, the arena's <em>owner thread</em>.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/Arena.java:244]: period?

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 4]
```diff
@@ -410,40 +463,49 @@ public sealed interface MemoryLayout permits SequenceLayout, GroupLayout, Paddin
...
      */
     long byteAlignment();
 
     /**
-     * {@return a memory layout with the same characteristics as this layout, but with the given
-     * alignment constraint (in bytes)}
+     * {@return a memory layout with the same characteristics as this layout, but with
+     *          the given alignment constraint (in bytes)}
      *
      * @param byteAlignment the layout alignment constraint, expressed in bytes.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:510]: period?

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 5]
```diff
@@ -865,11 +1012,12 @@ static PaddingLayout paddingLayout(long byteSize) {
     /**
      * Creates a sequence layout with the given element layout and element count.
      *
-     * @param elementCount the sequence element count.
-     * @param elementLayout the sequence element layout.
+     * @param elementCount the sequence element count
+     * @param elementLayout the sequence element layout
      * @return the new sequence layout with the given element layout and size.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:1017]: period?

File: src/java.base/share/classes/java/nio/channels/FileChannel.java [hunk 6]
```diff
@@ -1068,8 +1068,8 @@ public abstract MappedByteBuffer map(MapMode mode, long position, long size)
      *          If {@code arena.isAlive() == false}.
      *
      * @throws  WrongThreadException
-     *          If {@code arena} is a confined scoped arena, and this method is called from a
-     *          thread {@code T}, other than the scoped arena's owner thread.
+     *          If {@code arena} is a confined scoped arena, and this method is called
+     *          from a thread {@code T}, other than the scoped arena's owner thread.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/nio/channels/FileChannel.java:1072]: period?

**Review by mcimadamore — COMMENTED** [review]
Looks good. I found some (few) sentences that still have a full stop in `@param`/`@throws`

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 7]
```diff
@@ -916,8 +910,8 @@ static PathElement sequenceElement(long index) {
          * @param start the index of the first sequence element to be selected
          * @param step the step factor at which subsequence sequence elements are to be
          *             selected
-         * @return a path element which selects the sequence element layout with the
-         *         given index
+         * @return a path element which selects the sequence element layout in the
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:913]: Good catch!

**Review by mcimadamore — COMMENTED** [review]
Please create a specdiff between the old version and the version in this PR, so that we can take a look at the differences in the resulting javadoc, and make sure that nothing else got affected (by accident).

**Comment by minborg** [comment]
A specdiff is available here https://cr.openjdk.org/~pminborg/panama/22/specdiff-panamajavadocrework/package-summary.html

The specdiff is from this PR's latest commit (at the time of writing this message) and compares with ee57e731d03 which is a commit just before we started integrating a number of spec updates. This means the diff will include updates from other PRs including @JornVernee 's Linker improvements.

File: src/java.base/share/classes/java/lang/foreign/Arena.java [hunk 8]
```diff
@@ -31,47 +31,53 @@
 import java.lang.foreign.MemorySegment.Scope;
 
 /**
- * An arena controls the lifecycle of native memory segments, providing both flexible allocation and timely deallocation.
+ * An arena controls the lifecycle of native memory segments, providing both flexible
+ * allocation and timely de-allocation.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/Arena.java:35]: Please revert all uses of "de-allocation". I don't think that's incorrect - seems more of a pedantic check from the tool.

File: src/java.base/share/classes/java/lang/foreign/Linker.java [hunk 9]
```diff
@@ -759,8 +874,9 @@ static Option captureCallState(String... capturedState) {
         }
 
          /**
-         * {@return a struct layout that represents the layout of the capture state segment that is passed
-         *          to a downcall handle linked with {@link #captureCallState(String...)}}
+         * {@return a struct layout that represents the layout of the capture state
+          *         segment that is passedto a downcall handle linked with
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/Linker.java:878]: there's a mistake `passedto` w/o spaces

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 10]
```diff
@@ -407,12 +481,14 @@
...
+ * segment is known statically, the client can overlay a
+ * {@linkplain AddressLayout#withTargetLayout(MemoryLayout) target layout} on the address
  * layout used when reading a pointer. The target layout is then used to dynamically
- * <em>expand</em> the size of the native memory segment returned by the access operation, so that the size
- * of the segment is the same as the size of the target layout. In other words, the returned segment is no
- * longer a zero-length memory segment, and the pointer it represents can be dereferenced directly:
+ * <em>expand</em> the size of the native memory segment returned by the access operation
+ * so that the size of the segment is the same as the size of the target layout . In other
+ * words, the returned segment is no longer a zero-length memory segment, and the pointer
+ * it represents can be de-referenced directly:
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:491]: Please revert "de-referenced"

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 11]
```diff
@@ -1366,377 +1575,439 @@ static void copy(MemorySegment srcSegment, ValueLayout srcElementLayout, long sr
...
+     * layout size.
      *
-     * @param layout the layout of the region of memory to be read.
-     * @param index a logical index. The offset in bytes (relative to this segment address) at which the access operation
-     *              will occur can be expressed as {@code (index * layout.byteSize())}.
-     * @return a boolean value read from this segment.
-     * @throws IllegalStateException if the {@linkplain #scope() scope} associated with this segment is not
-     *         {@linkplain Scope#isAlive() alive}
+     * @param layout the layout of the region of memory to be read
+     * @param index a logical index, the offset in bytes (relative to this
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1956]: I think the new text for `index` is not good (here and elsewhere). I suggest:
```
a logical index _denoting_ the offset in bytes...
```

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 12]
```diff
@@ -1931,7 +1931,7 @@ static void copy(MemorySegment srcSegment, ValueLayout srcElementLayout, long sr
      * layout size.
      *
      * @param layout the layout of the region of memory to be read
-     * @param index a logical index, the offset in bytes (relative to this
+     * @param index a logical index denoting the offset in bytes (relative to this
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1934]: apologies for sending you the wrong way. As a single sentence this doesn't make sense. I suggest the original text (e.g. before this PR) is restored.

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**